debian-bo-master/bo/source/comm/term-2.3.5/termtest.1
.TH TERMTEST 1
.SH NAME
termtest \- debugging tool for 
.IR term
.SH SYNOPSIS
.B "termtest "
[\-1 <program>] [\-2 <program>] [-local <options>] [--remote <options>]
[<options>]
.SH DESCRIPTION
You should perform two sets of tests before you ever try to run
.IR term
itself.
The
.IR termtest
program lets you 
exercise 
.IR term
and its clients locally by starting two
.IR term
daemons on the same system.
.IR Termtest
when run, establishes two connected instances of 
.IR term 
on the local system so you may see if they and the clients work.  
To avoid problems with other programs also called 'termtest' you should 
run
.IR termtest 
from the source directory of 
.IR term
by typing
.PP
\ \ \ setenv TERMMODE 0
.PP
.PP
\ \ \ ./termtest \ \-n \ on
.PP
.IR Termtest
opens the files 'local.log' and 'remote.log' in the local directory to
record the stderr output of the two term daemons.
Options on the command line of
.B termtest
are passed on to both of them.
If you need to debug you can use 
.B \-d 64
or even
.B \-d 2112.
Familiarity with the sources is essential if you use debugging.
The termrc file is used to set the parameters of both daemons.
You can specify a different term daemon (e.g. an old one that is
already installed on your system) as the 'remote' by giving the
option 
.B \-2 program
as the first option for test. Similarly, you can use a
different 'local' term daemon with the 
.B \-1 
option.
.PP
Once 
.IR termtest
is running you should be able to use the clients from any shell:
.PP
\ \ \ trsh
.PP
give you a 'remote' shell.  You may need to type 'reset ^J' to reset
the tty of the new shell (that's Cntl-J).
.PP
\ \ \ trsh -s who
.PP
will run the who command 'remotely' to list the current users.
.PP
\ \ \ tupload -fvv <filename> ... /tmp
.PP
will copy the listed files into the /tmp directory and provide CPS statistics.
.PP
\ \ \ tredir 4000 23
.PP
will map port 4000 to 23 and put itself in the background.  Then, typing
.PP
\ \ \ telnet 0 4000
.PP
should give you a login prompt that works.
If you're running X, you should be able to use 
.IR txconn
to establish a new display (screen) that maps to the existing screen.
.PP
Be sure to run 
.IR ./termtest
on both the local and remote systems,
particularly if the systems use different architectures.
.SH BUGS
.SH SEE ALSO
.IR term (1),
.IR termrc (1),
.IR term_setup (1),
.IR term_clients (1),
.IR termtest (1),
.IR linecheck (1).
.SH AUTHOR
Michael O'Reilly, michael@iinet.com.au, original author
.br
Bill C. Riemers, bcr@physics.purdue.edu, current developer



debian-bo-master/bo/source/comm/term-2.3.5/termtest.c
/*  test.c - start two connected term daemons for testing purposes 

This program is for testing term. It starts two term daemons that are
connected. The "local" one listens on server (socket) name "" and
writes its output to "local.log". The "remote" one listens on server
name "remote", is started with the -r switch and writes its output to
"remote.log". This means that clients started without special
arguments connect to the "local" term daemon.

"test" understands the following options:
-1 program     - use "program"as the primary term daemon (default ./term)
-local option  - use "option" for the primary term daemon
-2 program     - use "option" as the secondary term daemon (default ./term)
-remote option - use "option" for the secondary term daemon

All other options will be given to both daemons.

This program prints the actual command lines that are started for both
daemons, then backgrounds itself. Killing either of its processes
or using "tshutdown" kills all of them. */

#define I_ERRNO
#define I_IOCTL
#define I_SIGNAL
#include "includes.h"

int childa=0, childb=0;

static void killall(int i) {
  kill(childa, SIGKILL);
  kill(childb, SIGKILL);
  exit(0);
}

int main(int argc, char *argv[]) {
    int i,j, use_defaults=1;
    int soc[2];
    char path[256];
    char *args[256];

#ifdef SOCKS
  SOCKSinit(argv[0]);
#endif
    S_Pipe(soc);

    if (!(childa = x__fork())) {
	sprintf(path, "%s/.term/socket", getenv("HOME"));
	unlink(path);
        args[(j=0)] = "./term";
        args[++j] = "-s";
        args[++j] = "off";
        for (i = 1; i < argc; ++i)
            if (!strcmp(argv[i],"-1")) {
              args[0] = argv[++i];
              j = -i;
              use_defaults = 0;
            }else if (!strcmp(argv[i],"--local")) {
              j--;
            }else if (!strcmp(argv[i],"--remote")||!strcmp(argv[i],"-2")) {
              i++;
              j -= 2;
            }else {
              args[i+j] = argv[i];
            }
        args[i+j] = NULL;
	sleep(1);
	x__dup2(soc[0], 0);
	x__dup2(soc[0], 1);
       	i = open("local.log", O_RDWR | O_CREAT | O_TRUNC , 0666);
	if (i < 0) {
	  x__perror("Open");
	  exit(1);
	}
	x__dup2(i, 2);
	fprintf(stderr,"Executing (%d): %s",getpid(),args[0]);
	for (i=1;args[i] != NULL; ++i) fprintf(stderr," %s",args[i]);
        fprintf(stderr,"\n");
	sleep(1);
	execv(args[0], args);
	x__perror("local execv failed");
	exit(1);
	}
    if (!(childb=x__fork())) {
	sprintf(path, "%s/.term/sockettest", getenv("HOME"));
	unlink(path);
        args[(j=0)] = "./term";
        args[++j] = "-r";
        args[++j] = "-s";
        args[++j] = "off";
        for (i = 1; i < argc; ++i)
            if (!strcmp(argv[i],"-2")) {
              args[0] = argv[++i];
              j = -i;
	      use_defaults = 0;
            }else if (!strcmp(argv[i],"--remote")) {
              j--;
            }else if (!strcmp(argv[i],"--local")||!strcmp(argv[i],"-1")) {
              i++;
              j -= 2;
            }else {
              args[i+j] = argv[i];
            }
        if (use_defaults) args[i+(j++)] = "test";
        args[i+j] = NULL;
	x__dup2(soc[1], 0);
	x__dup2(soc[1], 1);
       	i = open("remote.log", O_RDWR | O_CREAT | O_TRUNC , 0666);
	if (i < 0) {
	  x__perror("Open");
	  exit(1);
	}
	x__dup2(i, 2);
	fprintf(stderr,"Executing (%d): %s",getpid(),args[0]);
	for (i=1;args[i] != NULL; ++i) fprintf(stderr," %s",args[i]);
        fprintf(stderr,"\n");
	execv(args[0], args);
	x__perror("remote execv failed");
	exit(1);
	}
    signal(SIGCHLD, killall);
    signal(SIGTERM, killall);
    signal(SIGHUP, killall);
    signal(SIGQUIT, killall);
    signal(SIGTERM, killall);
    pause();
    killall(0);
    exit(0);
}
	
